### ISSUE TITLE
CrunchStatistics always fails

### ISSUE THREAD

**Issue by npalmqvist** [rinvex/laravel-statistics#11]
```Illuminate\Queue\MaxAttemptsExceededException: Rinvex\Statistics\Jobs\CrunchStatistics has been attempted too many times or run too long. The job may have previously timed out. in /Users/niklaspalmqvist/git-projects/HandbollOnlineAPI-v2/vendor/laravel/framework/src/Illuminate/Queue/Worker.php:401```

I have set my queue workers to a very long timeout but still only get this error from Horizon. 
Only have a few rows in my `statistics_data` table.

Have everything set up as in docs and scheduling the command like 
`$schedule->job(new \Rinvex\Statistics\Jobs\CrunchStatistics, 'data-worker')->everyHour();`

Grateful for help.

**Comment by robertgarrigos** [comment]
I'm having sort of the same problem. Just using the script without any job scheduling, only default lottery feature, but data never gets crunched.

**Comment by robertgarrigos** [comment]
I managed to get this to work, apparently, but nothing happens. I had a permissions problem logged. When fixed, and trying to run this command:

```
php artisan schedule:run
```
I got this result:
```
Running scheduled command: Rinvex\Statistics\Jobs\CrunchStatistics
Running scheduled command: Rinvex\Statistics\Jobs\CleanStatisticsRequests
```
which seems to be right. However, my data doesn't get crunched.

**Comment by robertgarrigos** [comment]
Right. Tracking this down, I added some logging functionality to CrunchStatistics.php, as the catch clause wasn't registering anything, and got this error logged when running the scheduled jobs:

```
local.ERROR: Watson\Validating\ValidationException: The given data was invalid. in /app/vendor/watson/validating/src/ValidatingTrait.php:436
```
Now I'm stuck as I'm new to laravel. I can just see that this is related to model validation. Any help on this would be much appreciated.

**Comment by robertgarrigos** [comment]
This is the rest of the log message:
```
#0 /app/vendor/watson/validating/src/ValidatingObserver.php(56): Rinvex\Statistics\Models\Route->throwValidationException()
#1 /app/vendor/watson/validating/src/ValidatingObserver.php(20): Watson\Validating\ValidatingObserver->performValidation(Object(Rinvex\Statistics\Models\Route), 'saving')
#2 [internal function]: Watson\Validating\ValidatingObserver->saving(Object(Rinvex\Statistics\Models\Route))
#3 /app/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(366): call_user_func_array(Array, Array)
#4 /app/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(196): Illuminate\Events\Dispatcher->Illuminate\Events\{closure}('eloquent.saving...', Array)
#5 /app/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(169): Illuminate\Events\Dispatcher->dispatch('eloquent.saving...', Array, true)
#6 /app/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/HasEvents.php(188): Illuminate\Events\Dispatcher->until('eloquent.saving...', Object(Rinvex\Statistics\Models\Route))
#7 /app/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(647): Illuminate\Database\Eloquent\Model->fireModelEvent('saving')
#8 /app/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(420): Illuminate\Database\Eloquent\Model->save()
#9 /app/vendor/laravel/framework/src/Illuminate/Support/helpers.php(1122): Illuminate\Database\Eloquent\Builder->Illuminate\Database\Eloquent\{closure}(Object(Rinvex\Statistics\Models\Route))
#10 /app/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(421): tap(Object(Rinvex\Statistics\Models\Route), Object(Closure))
#11 /app/vendor/laravel/framework/src/Illuminate/Support/Traits/ForwardsCalls.php(23): Illuminate\Database\Eloquent\Builder->firstOrCreate(Array, Array)
#12 /app/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1614): Illuminate\Database\Eloquent\Model->forwardCallTo(Object(Rinvex\Cacheable\EloquentBuilder), 'firstOrCreate', Array)
#13 /app/vendor/rinvex/laravel-statistics/src/Jobs/CrunchStatistics.php(54): Illuminate\Database\Eloquent\Model->__call('firstOrCreate', Array)
#14 /app/vendor/laravel/framework/src/Illuminate/Database/Concerns/BuildsQueries.php(62): Rinvex\Statistics\Jobs\CrunchStatistics->Rinvex\Statistics\Jobs\{closure}(Object(Rinvex\Statistics\Models\Datum), 4)
#15 /app/vendor/laravel/framework/src/Illuminate/Database/Concerns/BuildsQueries.php(39): Illuminate\Database\Eloquent\Builder->Illuminate\Database\Concerns\{closure}(Object(Illuminate\Database\Eloquent\Collection), 1)
#16 /app/vendor/laravel/framework/src/Illuminate/Database/Concerns/BuildsQueries.php(66): Illuminate\Database\Eloquent\Builder->chunk(1000, Object(Closure))
#17 /app/vendor/laravel/framework/src/Illuminate/Support/Traits/ForwardsCalls.php(23): Illuminate\Database\Eloquent\Builder->each(Object(Closure))
#18 /app/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1614): Illuminate\Database\Eloquent\Model->forwardCallTo(Object(Rinvex\Cacheable\EloquentBuilder), 'each', Array)
#19 /app/vendor/rinvex/laravel-statistics/src/Jobs/CrunchStatistics.php(130): Illuminate\Database\Eloquent\Model->__call('each', Array)
#20 [internal function]: Rinvex\Statistics\Jobs\CrunchStatistics->handle()
#21 /app/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(32): call_user_func_array(Array, Array)
#22 /app/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(90): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#23 /app/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(34): Illuminate\Container\BoundMethod::callBoundMethod(Object(Illuminate\Foundation\Application), Array, Object(Closure))
#24 /app/vendor/laravel/framework/src/Illuminate/Container/Container.php(580): Illuminate\Container\BoundMethod::call(Object(Illuminate\Foundation\Application), Array, Array, NULL)
#25 /app/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(94): Illuminate\Container\Container->call(Array)
#26 /app/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\Bus\Dispatcher->Illuminate\Bus\{closure}(Object(Rinvex\Statistics\Jobs\CrunchStatistics))
#27 /app/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(104): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Rinvex\Statistics\Jobs\CrunchStatistics))
#28 /app/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(98): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#29 /app/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(49): Illuminate\Bus\Dispatcher->dispatchNow(Object(Rinvex\Statistics\Jobs\CrunchStatistics), false)
#30 /app/vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php(86): Illuminate\Queue\CallQueuedHandler->call(Object(Illuminate\Queue\Jobs\SyncJob), Array)
#31 /app/vendor/laravel/framework/src/Illuminate/Queue/SyncQueue.php(42): Illuminate\Queue\Jobs\Job->fire()
#32 /app/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(184): Illuminate\Queue\SyncQueue->push(Object(Rinvex\Statistics\Jobs\CrunchStatistics))
#33 /app/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(160): Illuminate\Bus\Dispatcher->pushCommandToQueue(Object(Illuminate\Queue\SyncQueue), Object(Rinvex\Statistics\Jobs\CrunchStatistics))
#34 /app/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(73): Illuminate\Bus\Dispatcher->dispatchToQueue(Object(Rinvex\Statistics\Jobs\CrunchStatistics))
#35 /app/vendor/laravel/framework/src/Illuminate/Foundation/Bus/PendingDispatch.php(112): Illuminate\Bus\Dispatcher->dispatch(Object(Rinvex\Statistics\Jobs\CrunchStatistics))
#36 /app/vendor/laravel/framework/src/Illuminate/Console/Scheduling/Schedule.php(112): Illuminate\Foundation\Bus\PendingDispatch->__destruct()
#37 [internal function]: Illuminate\Console\Scheduling\Schedule->Illuminate\Console\Scheduling\{closure}()
#38 [internal function]: Closure->__invoke()
#39 /app/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(32): call_user_func_array(Array, Array)
#40 /app/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(90): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#41 /app/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(34): Illuminate\Container\BoundMethod::callBoundMethod(Object(Illuminate\Foundation\Application), Array, Object(Closure))
#42 /app/vendor/laravel/framework/src/Illuminate/Container/Container.php(580): Illuminate\Container\BoundMethod::call(Object(Illuminate\Foundation\Application), Array, Array, NULL)
#43 /app/vendor/laravel/framework/src/Illuminate/Console/Scheduling/CallbackEvent.php(75): Illuminate\Container\Container->call(Array, Array)
#44 /app/vendor/laravel/framework/src/Illuminate/Console/Scheduling/ScheduleRunCommand.php(111): Illuminate\Console\Scheduling\CallbackEvent->run(Object(Illuminate\Foundation\Application))
#45 /app/vendor/laravel/framework/src/Illuminate/Console/Scheduling/ScheduleRunCommand.php(75): Illuminate\Console\Scheduling\ScheduleRunCommand->runEvent(Object(Illuminate\Console\Scheduling\CallbackEvent))
#46 [internal function]: Illuminate\Console\Scheduling\ScheduleRunCommand->handle()
#47 /app/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(32): call_user_func_array(Array, Array)
#48 /app/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(90): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#49 /app/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(34): Illuminate\Container\BoundMethod::callBoundMethod(Object(Illuminate\Foundation\Application), Array, Object(Closure))
#50 /app/vendor/laravel/framework/src/Illuminate/Container/Container.php(580): Illuminate\Container\BoundMethod::call(Object(Illuminate\Foundation\Application), Array, Array, NULL)
#51 /app/vendor/laravel/framework/src/Illuminate/Console/Command.php(183): Illuminate\Container\Container->call(Array)
#52 /app/vendor/symfony/console/Command/Command.php(255): Illuminate\Console\Command->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Illuminate\Console\OutputStyle))
#53 /app/vendor/laravel/framework/src/Illuminate/Console/Command.php(170): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Illuminate\Console\OutputStyle))
#54 /app/vendor/symfony/console/Application.php(908): Illuminate\Console\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#55 /app/vendor/symfony/console/Application.php(269): Symfony\Component\Console\Application->doRunCommand(Object(Illuminate\Console\Scheduling\ScheduleRunCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#56 /app/vendor/symfony/console/Application.php(145): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#57 /app/vendor/laravel/framework/src/Illuminate/Console/Application.php(90): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#58 /app/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(122): Illuminate\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#59 /app/artisan(37): Illuminate\Foundation\Console\Kernel->handle(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#60 {main}  
```

**Comment by pawel-marciniak** [comment]
Hey, what version of Laravel are you using? My data crunching also stopped :/ It seems like it stopped after update of something some time ago.

**Comment by ermyril** [comment]
Well, I'm using laravel 7.4.0 along with php 7.4, in docker. And experiencing the same issue sadly. Don't really know where to look

**Comment by genesiscz** [comment]
How is this still open?

This code is forcing to create a route while most routes have NULL as a name, making validation tofail 

```
                $route = app('rinvex.statistics.route')->firstOrCreate([
                    'name' => $laravelRoute->getName(),
                ], [
                    'path' => $laravelRoute->uri(),
                    'action' => $laravelRoute->getActionName(),
                    'middleware' => $laravelRoute->gatherMiddleware() ?: null,
                    'parameters' => $tokens ?: null,
                ]);
```

**Comment by genesiscz** [comment]
- I cleared "required" rules in Route model
- Changed above code to

```
                $route = app('rinvex.statistics.route')->firstOrCreate([
                    'action' => $laravelRoute->getActionName(),
                    'name' => $laravelRoute->getName() ?? "",
                ], [
                    'path' => $laravelRoute->uri(),
                    'middleware' => $laravelRoute->gatherMiddleware() ?: null,
                    'parameters' => $tokens ?: null,
                ]);
```

- Changed unique index to be from action,name
- It works

**Comment by kondratovbr** [comment]
Weirdly, for me CrunchStatistics executes but no counters get updated and no exceptions are thrown.

**Comment by genesiscz** [comment]
> Weirdly, for me CrunchStatistics executes but no counters get updated and no exceptions are thrown.

Because the command has a big try catch(Exception) with no followup. Try to dig into whereabouts the code i posted and remove it from the try block

**Comment by Omranic** [comment]
Thank you for your patience, this has been fixed in #19

**Comment by jrbecart** [comment]
I had the same error with getName but also with gatherMiddleware on some other routes. 
When calling gatherMiddleware() with my php-worker the job timeout, but when running it manually it took more than 2:30min 
Adjusting your worker timeout should work, but I have no idea why it's taking so long.
